File: src/api/schemas/articles.py
"""Request/response schemas for the articles API."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class GenerateArticleRequest(BaseModel):
    session_id: UUID


class OutlineSectionResponse(BaseModel):
    index: int
    title: str
    description: str
    key_points: list[str]
    target_word_count: int
    relevant_facets: list[int]


class ArticleOutlineResponse(BaseModel):
    draft_id: UUID
    title: str
    subtitle: str | None
    content_type: str
    sections: list[OutlineSectionResponse]
    total_target_words: int
    reasoning: str
    status: str


class CitationRefResponse(BaseModel):
    index: int
    source_url: str
    source_title: str


class SectionDraftResponse(BaseModel):
    section_index: int
    title: str
    body_markdown: str
    word_count: int
    citations_used: list[CitationRefResponse]


class StructuredDataLDResponse(BaseModel):
    headline: str
    description: str
    keywords: list[str]
    date_published: str
    date_modified: str


class SEOResultResponse(BaseModel):
    title: str
    description: str
    keywords: list[str]
    summary: str
    key_claims: list[str]
    ai_disclosure: str
    structured_data: StructuredDataLDResponse | None = None


class ArticleDraftResponse(BaseModel):
    draft_id: UUID
    session_id: UUID
    status: str
    outline: ArticleOutlineResponse | None
    created_at: datetime
    completed_at: datetime | None
    section_drafts: list[SectionDraftResponse] = []
    citations: list[CitationRefResponse] = []
    total_word_count: int = 0
    seo_result: SEOResultResponse | None = None


class CitationResponse(BaseModel):
    index: int
    title: str
    url: str
    authors: list[str]
    published_at: datetime | None


class ProvenanceResponse(BaseModel):
    research_session_id: UUID
    primary_model: str
    drafting_model: str
    embedding_model: str
    embedding_version: str


class ImageAssetResponse(BaseModel):
    id: UUID
    url: str
    caption: str | None
    alt_text: str | None


class SEOMetadataResponse(BaseModel):
    title: str
    description: str
    keywords: list[str]
    canonical_url: str | None
    structured_data: StructuredDataLDResponse | None = None


class CanonicalArticleResponse(BaseModel):
    id: UUID
    title: str
    subtitle: str | None
    body_markdown: str
    summary: str
    key_claims: list[str]
    content_type: str
    seo: SEOMetadataResponse
    citations: list[CitationResponse]
    visuals: list[ImageAssetResponse]
    authors: list[str]
    domain: str
    generated_at: datetime
    provenance: ProvenanceResponse
    ai_generated: bool

File: frontend/src/types/articles.ts
// AUTHOR-007 — real editorial states ("complete" removed; the badge keeps
// a legacy alias for the dashboard's separate Article union in types/api.ts).
export type ArticleStatus = "draft" | "in_review" | "approved" | "published";

export interface ArticleListItem {
  id: string;
  title: string;
  summary: string;
  domain: string;
  status: ArticleStatus;
  wordCount: number;
  generatedAt: string;
}

export interface Citation {
  index: number;
  title: string;
  url: string;
  authors: string[];
  publishedAt: string | null;
}

export interface Provenance {
  researchSessionId: string;
  primaryModel: string;
  draftingModel: string;
  embeddingModel: string;
  embeddingVersion: string;
}

export interface StructuredDataLD {
  headline: string;
  description: string;
  keywords: string[];
  datePublished: string;
  dateModified: string;
}

export interface SEOMetadata {
  title: string;
  description: string;
  keywords: string[];
  canonicalUrl: string | null;
  structuredData: StructuredDataLD | null;
}

export interface ImageAssetMetadata {
  diagram_type?: string;
  source_section?: number;
  mermaid_syntax?: string;
  // Image-planner (Epic 10) fields. The planner stamps `section_index`
  // (not the legacy `source_section`), plus the placement anchor and the
  // role_style used to decide cover vs inline rendering.
  section_index?: number;
  placement_anchor?: string;
  role_style?: string;
}

export interface ImageAsset {
  id: string;
  url: string;
  caption: string | null;
  altText: string | null;
  metadata?: ImageAssetMetadata | null;
}

export interface WorkflowStep {
  name: string;
  durationSeconds: number;
}

export interface ArticleDetail {
  id: string;
  title: string;
  subtitle: string | null;
  bodyMarkdown: string;
  summary: string;
  keyClaims: string[];
  contentType: string;
  seo: SEOMetadata;
  citations: Citation[];
  visuals: ImageAsset[];
  authors: string[];
  domain: string;
  generatedAt: string;
  provenance: Provenance;
  aiGenerated: boolean;
  status: ArticleStatus;
  wordCount: number;
  workflow: WorkflowStep[];
}

// AUTHOR-006 — metadata editing wire types
export interface ArticleMetadataPatch {
  title?: string;
  subtitle?: string;
  seo_title?: string;
  seo_description?: string;
  keywords?: string[];
  status?: ArticleStatus; // AUTHOR-007
}

export interface FieldWarning {
  field: string;
  message: string;
}

export interface ArticleMetadataResult {
  id: string;
  title: string;
  subtitle: string | null;
  seo: { title: string; description: string; keywords: string[] };
  warnings: FieldWarning[];
}

export type SeoRegenerateField = "seo_title" | "seo_description" | "keywords";

export interface SeoRegenerateResult {
  field: string;
  value: string | string[];
  warnings: FieldWarning[];
}
